contracts/oracles/v2/adapters/balancer/BalancerV2PriceOracle.sol

```diff
@@ -0,0 +1,197 @@
+    mapping(address => address) public tokenPools; // token => balancer pool
+
+    event SetTokenOracle(
+        address indexed token,
+        address indexed tokenPool,
+        address indexed denominatedOracle
+    );
+
+    /**
+     * @dev sets up the Price Oracle
+     * @param _roles exodia roles address
+     * @param _vault balancer vault address
+     * @param _minimumUpdateInterval how often to permit updates to the TWAP (seconds)
+     *                               If set to 0, will use the default of 5 minutes
+     */
+    function initialize(
+        address _roles,
+        address _vault,
+        uint256 _minimumUpdateInterval
+    ) public initializer {
+        require(_roles != address(0), "roles cannot be null address");
+        require(_vault != address(0), "vault cannot be null address");
+
+        __ExodiaAccessControl__init(_roles);
+
+        vault = _vault;
+        if (_minimumUpdateInterval != 0) {
+            minimumUpdateInterval = _minimumUpdateInterval;
+        }
+    }
+
+    /**
+     * @dev add/update token oracle setting
+     */
+    function setTokenOracle(
+        address _token,
+        address _tokenPool,
+        address _denominatedOracle
+    ) external onlyArchitect {
+        bytes32 poolId = IBPoolV2(_tokenPool).getPoolId();
+        (IERC20[] memory tokens, , ) = IBVaultV2(vault).getPoolTokens(poolId);
+
+        require(tokens.length == 2, "INVALID POOL");
+        require(
+            _token == address(tokens[0]) || _token == address(tokens[1]),
+            "INVALID TOKENS"
+        );
+
+        denominatedOracles[_token] = _denominatedOracle;
+        tokenPools[_token] = _tokenPool;
+
+        emit SetTokenOracle(_token, _tokenPool, _denominatedOracle);
+    }
+
+    /****** OPERATIONAL METHODS ******/
+
+    /**
+     * @dev returns the TWAP for the provided pair as of the last update
+     */
+    function getSafePrice(address _token) public view returns (uint256 price) {
+        require(
+            tokenPools[_token] != address(0) && denominatedOracles[_token] != address(0),
+            "UNSUPPORTED"
+        );
+
+        IBalV2PriceOracle.OracleAverageQuery[]
+            memory query = new IBalV2PriceOracle.OracleAverageQuery[](1);
+        query[0] = IBalV2PriceOracle.OracleAverageQuery(
+            IBalV2PriceOracle.TWAP_VALUE.PAIR_PRICE,
+            minimumUpdateInterval,
+            10
+        );
+        uint256[] memory prices = IBalV2PriceOracle(tokenPools[_token])
+            .getTimeWeightedAverage(query);
+        uint256 tokenPairPrice = prices[0];
+
+        bytes32 poolId = IBPoolV2(tokenPools[_token]).getPoolId();
+        (IERC20[] memory tokens, , ) = IBVaultV2(vault).getPoolTokens(poolId);
+
+        if (_token == address(tokens[0])) {
+            price =
+                (_denominatedTokenPrice(
+                    IPriceOracle(denominatedOracles[_token]),
```

Comment:
You are casting the address to an IPriceOracle in the function, I would remove this cast here